### PR TITLE
New feature implementation of writing AAF files from otio files in the AAF adapter

### PR DESCRIPTION
**Pull request by freesonluxo**
This is an initial implementation of the AAF writer in AAF adapter.

 Co-authored-by: Shahbaz Khan <shahbazkhan8194@gmail.com>

### CODE REVIEW

**Comment by shahbazk8194** [comment]
This is ready for a review!

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+            slot_id=self.timeline_mobslot.slot_id,
+            length=length,
+            media_kind=self.media_kind)
+        compmob_clip.mob = mastermob
+        compmob_clip.slot = mastermob_slot
+        compmob_clip.slot_id = mastermob_slot.slot_id
+        return compmob_clip
+
+    def aaf_transition(self, otio_transition):
+        """Convert an otio Transition into an aaf Transition."""
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Check if this is a cross dissolve

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+        c1.time = pointlist[0]["Time"]
+
+        c2 = self.f.create.ControlPoint()
+        c2["EditHint"].value = "Proportional"
+        c2.value = pointlist[1]["Value"]
+        c2.time = pointlist[1]["Time"]
+
+        opacity_u["PointList"].extend([c1, c2])
+
+        op_group_metadata = otio_transition.metadata["AAF"]["OperationGroup"]
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Think about default values...

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+            timecode_length = _timecode_length(otio_clip)
+            tape_timecode_slot.segment.start = timecode_start
+            tape_timecode_slot.segment.length = timecode_length
+            self.f.content.mobs.append(tape_mob)
+            self._unique_tapemobs[mob_id] = tape_mob
+        return tape_mob
+
+    def track_transcriber(self, otio_track):
+        """Return an appropriate TrackTranscriber given an otio track."""
+        if otio_track.kind == "Video":
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Use enum

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+            transcriber = VideoTrackTranscriber(self, otio_track)
+        elif otio_track.kind == "Audio":
+            transcriber = AudioTrackTranscriber(self, otio_track)
+        assert transcriber
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Delete this.

Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Raise exception

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+        errors = []
+
+        def _check(otio_child, keys_path):
+            keys = keys_path.split(".")
+            value = otio_child.metadata
+            try:
+                for key in keys:
+                    value = value[key]
+            except KeyError:
+                print "{}({}) is missing required metadata {}".format(
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: pytion3-ified

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+                    otio_child.name, type(otio_child), keys_path)
+                errors.append((otio_child, keys_path))
+
+        for otio_track in timeline.tracks:
+            for otio_child in otio_track:
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: each_child()

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+        """Return an appropriate TrackTranscriber given an otio track."""
+        if otio_track.kind == "Video":
+            transcriber = VideoTrackTranscriber(self, otio_track)
+        elif otio_track.kind == "Audio":
+            transcriber = AudioTrackTranscriber(self, otio_track)
+        assert transcriber
+        return transcriber
+
+    @staticmethod
+    def precheck(timeline):
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: validate_metadata

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+    def track_transcriber(self, otio_track):
+        """Return an appropriate TrackTranscriber given an otio track."""
+        if otio_track.kind == "Video":
+            transcriber = VideoTrackTranscriber(self, otio_track)
+        elif otio_track.kind == "Audio":
+            transcriber = AudioTrackTranscriber(self, otio_track)
+        assert transcriber
+        return transcriber
+
+    @staticmethod
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: move out to a free function

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 9]
```diff
@@ -0,0 +1,487 @@
+#
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: aaf_writer.py ?

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+
+class TrackTranscriber(object):
+    """
+    TrackTranscriber is the base class for the conversion of a given otio track.
+
+    TrackTranscriber is not meant to be used by itself. It provides the common
+    functionality to inherit from.
+    """
+
+    def __init__(self, aaf_file_transcriber, otio_track):
```
Comment (freesonluxo) [inline opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: parent_transcriber

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py (hunk 1 above)
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: @shahbazk8194 will look into this

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py (hunk 2 above)
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: @shahbazk8194 will look into this

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py (hunk 10 above)
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Rename this to parent_transcriber.

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+        filemob_clip.slot_id = tapemob_slot.slot_id
+        filemob_slot.segment = filemob_clip
+        return filemob, filemob_slot
+
+    def _create_timeline_mobslot(self):
+        # TimelineMobSlot
+        timeline_mobslot = self.compositionmob.create_sound_slot(
+            edit_rate=self.edit_rate)
+        # OperationDefinition
+        opdef_auid = AUID("9d2ea893-0968-11d3-8a38-0050040ef7d2")
```
[the remark this answers is not in the document]
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: Need to ping Mark to add accessible Enums

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+
+    @property
+    def media_kind(self):
+        """Return the string for what kind of track this is."""
+        raise NotImplementedError()
+
+    def aaf_filler(self, otio_gap):
+        """Convert an otio Gap into an aaf Filler"""
+        # length = otio_gap.duration().value  # XXX Not working for some reason
+        length = otio_gap.metadata["AAF"]["Length"]
```
[the remark this answers is not in the document]
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: We created an issue for this:
https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/438

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+
+
+def _timecode_length(clip):
+    """Return the timecode length (length of essence data) from an otio clip."""
+    try:
+        timecode_length = clip.media_reference.available_range.duration.value
+    except AttributeError:
+        timecode_length = clip.metadata["AAF"]["Length"]
+        print("WARNING: FIX ME! Relying on backup metadata value ('AAF.Length')"
+              "instead of actual media_reference.available_range.duration")
```
[the remark this answers is not in the document]
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: see https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/438

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+        Returns:
+            Returns a tuple of (MasterMob, MasterMobSlot)
+        """
+        mastermob = self.aaf_file_transcriber._unique_mastermob(otio_clip)
+        timecode_length = _timecode_length(otio_clip)
+        slot_id = otio_clip.metadata["AAF"]["SourceMobSlotID"]
+        try:
+            mastermob_slot = mastermob.slot_at(slot_id)
+        except IndexError:
+            mastermob_slot = mastermob.create_timeline_slot(
```
[the remark this answers is not in the document]
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: see https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/439

File: opentimelineio_contrib/adapters/aaf_adapter/transcriber.py (hunk 2 above)
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/transcriber.py]: see https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/439

File: opentimelineio_contrib/adapters/aaf_adapter/aaf_writer.py [hunk 15]
```diff
@@ -0,0 +1,508 @@
...
+        """
+        pass
+
+    @abstractmethod
+    def default_descriptor(self, otio_clip):
+        pass
+
+    def aaf_filler(self, otio_gap):
+        """Convert an otio Gap into an aaf Filler"""
+        # length = otio_gap.duration().value  # XXX Not working for some reason
```
[the remark this answers is not in the document]
Comment (freesonluxo) [reply opentimelineio_contrib/adapters/aaf_adapter/aaf_writer.py]: We filed this bug
https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/438
Right now we don't know why duration() isn't working. It requires investigation

**Pull request merged**